UsingCombineTests/MergingPipelineTests.swift

```diff
@@ -0,0 +1,87 @@
+
+        // set up the inputs and timing
+        let testablePublisher1: TestablePublisher<String, Never> = testScheduler.createRelativeTestablePublisher([
+            (100, .input("a")),
+            (200, .input("b")),
+            (350, .input("c")),
+        ])
+        let testablePublisher2: TestablePublisher<Int, Never> = testScheduler.createRelativeTestablePublisher([
+            (100, .input(1)),
+            (250, .input(2)),
+            (300, .input(3)),
+        ])
+
+        let merged = Publishers.CombineLatest(testablePublisher1, testablePublisher2)
+
+        // validate
+
+        // run the virtual time scheduler
+        let testableSubscriber = testScheduler.start { return merged }
+
+        // check the collected results
+        XCTAssertEqual(testableSubscriber.recordedOutput.count, 6)
+
+        print(testableSubscriber.recordedOutput)
+        // TestSequence<(String, Int), Never>(contents: [(200, .subscribe), (300, .input(("a", 1))), (400, .input(("b", 1))), (450, .input(("b", 2))), (500, .input(("b", 3))), (550, .input(("c", 3)))])
+
+        let firstInSequence = testableSubscriber.recordedOutput[0]
+        XCTAssertEqual(firstInSequence.0, 200) // checks the virtualtime of the subscription
+        // which is always expected to be 200 with this scheduled
+
+        // filter the output signals down to just the inputs - drop any subscriptions, cancel, or completions
+        let outputSignals = testableSubscriber.recordedOutput.filter { time, signal -> Bool in
+            // input type is (VirtualTime, Signal<(String, Int), Never>)
+            switch signal {
+            case .input(_, _):
+                return true
+            default:
+                return false
+            }
+        }
+        XCTAssertEqual(outputSignals.count, 5)
+        print(outputSignals)
+        // TestSequence<(String, Int), Never>(contents: [(300, .input(("a", 1))), (400, .input(("b", 1))), (450, .input(("b", 2))), (500, .input(("b", 3))), (550, .input(("c", 3)))])
+
+//        let foo = outputSignals[0]
+        //XCTAssertEqual(foo, .input(("a", 1))) // not equatable, so the compiler blows this up...
+        /*
+         Global function 'XCTAssertEqual(_:_:_:file:line:)' requires that '(VirtualTime, Signal<(String, Int), Never>)' conform to 'Equatable'
+         */
+
+        let expectedSequence = TestSequence<(String, Int), Never>([
+            (300, .input(("a", 1))),
+            (400, .input(("b", 1))),
+            (450, .input(("b", 2))),
+            (500, .input(("b", 3))),
+            (550, .input(("c", 3)))
+        ])
+
+        // XCTAssertEqual(outputSignals, expectedSequence) // compiler error
+        /*
+         Global function 'XCTAssertEqual(_:_:_:file:line:)' requires that '(String, Int)' conform to 'Equatable'
+         -> Requirement from conditional conformance of 'TestSequence<(String, Int), Never>' to 'Equatable' (EntwineTest.TestSequence<τ_0_0, τ_0_1>)
+         */
```

Comment:
@tcldr 👋 If you have a moment, I'm struggling here to figure out how to "pull apart" the results of TestSequence to assert an expected match when the embedded type is a combine tuple. In this case, it's <(String, Int)> - the result of using combineLatest on two separate simpler types.

I tried to two approaches, neither of which was entirely successful.

The first was to pull apart the individual element by digging into the types, tuples, and enums to get to specific values.
The second was to attempt to assemble the correct type and then try to use equatable TestSequences to validate things.
In both cases, it's conformance to equatable, where the type is a tuple or better of complexity, and there's no clear way for the compiler to understand how to say "yep, this is equal" or not.

Do you have any suggestions on how to tackle this kind of problem? It's not really an issue with EntwineTest, but you seemed pretty familiar with swift generics and enums and using them effectively, so I was hoping you might be able to provide a pointer or two.

Reply:
Leaving this here as well:

Tuples in Swift still leave a little to be desired, they don't auto-synthesise Equatable or Hashable conformance, and in fact there is no generic way you can conform a tuple to these protocols. From the discussions I've read from the Swift team, tuples often add an additional layer of complication when they're implementing new features and seem to be a general thorn in their sides!

However, they are particularly ergonomic from a coding perspective, so hopefully they'll resolve these issues at some point.

With EntwineTest, the solution I use to make the TestSequence conform to Equatable is to map the tuples to an intermediate struct, I then make _that_ struct conform to Equatable and compare those instead.

You could generalise it, but as I'm only using it in the one place I didn't want to bloat the codebase.

But it's simple enough to do.

The theory behind it is that any tuple can also be represented as a struct. So if you can create an intermediate struct that maps from a tuple and back, you then should have all you need to conform to whatever behaviours you need.

The big thing to remember is that you will need an intermediate struct for each arity of tuple. So for two, three, and four element tuples, you will need the intermediate structs of:

```
Tuple2<T0, T1>
Tuple3<T0, T1, T2>
Tuple4<T0, T1, T2, T3>
```
In terms of implementation it's as simple as:

```
struct Tuple2<T0, T1> {
    
    let t0: T0
    let t1: T1
    
    init(_ tuple: (T0, T1)) {
        self.t0 = tuple.0
        self.t1 = tuple.1
    }
    
    var raw: (T0, T1) { (t0, t1) }
}
```

And then Equatable and Hashable conformance is just:

```
extension Tuple2: Equatable where T0: Equatable, T1: Equatable {}
extension Tuple2: Hashable where T0: Hashable, T1: Hashable {}
```

Specifically with Entwine, though I think the generalised Tuples solution is probably outside the scope, what would probably help is some additional mapping functions so If you have made some intermediate structs – you can use them with ease.

Enabling something like:

```
    func testTuples() {
        
        let testScheduler = TestScheduler(initialClock: 0)
        
        let testablePublisher: TestablePublisher<(Int, String), Never> = testScheduler.createAbsoluteTestablePublisher([
            (  0, .input((0, "A"))),
            (200, .input((0, "B"))),
            (400, .input((0, "C"))),
        ])
        
        let testableSubscriber = testScheduler.start { testablePublisher }
        
        let results = testableSubscriber.recordedOutput
        let expected: TestSequence<(Int, String), Never> = [
            (200, .subscription),
            (200, .input((0, "B"))),
            (400, .input((0, "C"))),
        ]
        
        XCTAssertEqual(results.mapInput(Tuple2.init), expected.mapInput(Tuple2.init))
    }
```

Where Tuple2 is the struct above.